src/lxc/network.c

```diff
@@ -2679,6 +2717,151 @@ bool lxc_delete_network_unpriv(struct lxc_handler *handler)
+	char cmd_output[PATH_MAX];
+	struct ip_proxy_args args;
+	args.ip = ip;
+	args.dev = dev;
+
+	ret = run_command(cmd_output, sizeof(cmd_output),
+			  lxc_add_ip_proxy_exec_wrapper, (void *)&args);
+	if (ret < 0) {
+		ERROR("Failed to add ip proxy \"%s\" to dev \"%s\": %s", ip, dev, cmd_output);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int lxc_del_ip_proxy(const char *ip, const char *dev)
+{
+	int ret;
+	char cmd_output[PATH_MAX];
+	struct ip_proxy_args args;
+	args.ip = ip;
+	args.dev = dev;
+
+	ret = run_command(cmd_output, sizeof(cmd_output),
+			  lxc_del_ip_proxy_exec_wrapper, (void *)&args);
+	if (ret < 0) {
+		ERROR("Failed to delete ip proxy \"%s\" to dev \"%s\": %s", ip, dev, cmd_output);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int lxc_setup_l2proxy(struct lxc_netdev *netdev) {
+	struct lxc_list *cur, *next;
+	struct lxc_inetdev *inet4dev;
+	struct lxc_inet6dev *inet6dev;
+	char bufinet4[INET_ADDRSTRLEN], bufinet6[INET6_ADDRSTRLEN];
+
+	/* If IPv6 addresses are specified, then check that sysctl is configured correctly. */
+	if (!lxc_list_empty(&netdev->ipv6)) {
+		/* Check for net.ipv6.conf.[link].proxy_ndp=1 */
+		if (lxc_is_ip_neigh_proxy_enabled(netdev->link, AF_INET6) < 0) {
+			ERROR("Requires sysctl net.ipv6.conf.%s.proxy_ndp be set to 1", netdev->link);
+			return minus_one_set_errno(EINVAL);
+		}
+
+		/* Check for net.ipv6.conf.[link].forwarding=1 */
+		if (lxc_is_ip_forwarding_enabled(netdev->link, AF_INET6) < 0) {
+			ERROR("Requires sysctl net.ipv6.conf.%s.forwarding be set to 1", netdev->link);
+			return minus_one_set_errno(EINVAL);
+		}
+	}
+
+	lxc_list_for_each_safe(cur, &netdev->ipv4, next) {
+		inet4dev = cur->elem;
+		if (!inet_ntop(AF_INET, &inet4dev->addr, bufinet4, sizeof(bufinet4))) {
+			return minus_one_set_errno(EINVAL);
+		}
+
+		if (lxc_add_ip_proxy(bufinet4, netdev->link)) {
+			return minus_one_set_errno(EINVAL);
+		}
+	}
+
+	lxc_list_for_each_safe(cur, &netdev->ipv6, next) {
+		inet6dev = cur->elem;
+		if (!inet_ntop(AF_INET6, &inet6dev->addr, bufinet6, sizeof(bufinet6))) {
+			return minus_one_set_errno(EINVAL);
+		}
+
+		if (lxc_add_ip_proxy(bufinet6, netdev->link)) {
+			return minus_one_set_errno(EINVAL);
+		}
+	}
+
+	return 0;
+}
+
+static int lxc_delete_l2proxy(struct lxc_netdev *netdev) {
+	struct lxc_list *cur, *next;
+	struct lxc_inetdev *inet4dev;
+	struct lxc_inet6dev *inet6dev;
+	char bufinet4[INET_ADDRSTRLEN], bufinet6[INET6_ADDRSTRLEN];
+	int err = 0;
+
+	lxc_list_for_each_safe(cur, &netdev->ipv4, next) {
+		inet4dev = cur->elem;
+		if (!inet_ntop(AF_INET, &inet4dev->addr, bufinet4,sizeof(bufinet4))) {
+			err = -1;
+			continue; /* Try to remove any other l2proxy entries */
+		}
+
+		if (lxc_del_ip_proxy(bufinet4, netdev->link)) {
+			err = -1;
+			continue; /* Try to remove any other l2proxy entries */
+		}
+	}
+
+	lxc_list_for_each_safe(cur, &netdev->ipv6, next) {
+		inet6dev = cur->elem;
+		if (!inet_ntop(AF_INET6, &inet6dev->addr, bufinet6, sizeof(bufinet6))) {
+			err = -1;
```

Comment:
Should probably log which ip proxy failed.

Reply:
See https://github.com/lxc/lxc/pull/2964/files#r280351891